src/pip/_internal/commands/lock.py

```diff
@@ -151,18 +152,23 @@ def run(self, options: Values, args: List[str]) -> int:
         requirement_set = resolver.resolve(reqs, check_supported_wheels=True)
 
         if options.output_file == "-":
-            base_dir = Path.cwd()
+            # The pylock spec says "If a relative path is used it MUST be
+            # relative to the location of this file". When writing to stdout, we
+            # don't have a location for the file, so we can't use a relative
+            # path.
+            base_dir = None
         else:
             output_file_path = Path(options.output_file)
-            if not is_valid_pylock_file_name(output_file_path):
+            if not is_valid_pylock_path(output_file_path):
                 logger.warning(
                     "%s is not a valid lock file name.",
                     output_file_path,
                 )
             base_dir = output_file_path.parent
```

Comment:
I’m thinking, in many situations it might not actually be desirable to use a relative path. Not sure if there’s a good way to tell which way the user wants each package to be. Since outputting to stdout always generates an absolute path, _maybe_ it’s less confusing if absolute path is the default when you output to a file too, and you need to somehow tell pip to use a relative path explicitly. Something similar to how we have `--only-binary` for install? I don’t know.

Reply:
I'm not sure either. I'd say relative paths *below* the `pylock.toml` location is a good default. And it gives a nice result for the common `pip lock -e .`.

Anyway, I'm going to remove this change from the PR and handle that separately.

Reply:
I feel as though when you generate a pylock *file*, it makes sense for the paths to be relative to the location of the file. But when you generate to stdout (which has no location), it makes sense for paths to be absolute. As you say, though, maybe we don't want relative paths that go *up* in the filesystem hierarchy.

I'd go with:

1. When writing to stdout, absolute.
2. When writing to pylock, if the target is alongside (or below) the lockfile, then relative, ohterwise absolute.

But it's debatable enough that I think doing it as a separate PR is sensible.